File: lindera-python/tests/test_argument_depth.py
"""Tests for the nesting limit on converted arguments (lindera/lindera#1062)."""

import subprocess
import sys
import textwrap

import pytest

from lindera import TokenizerBuilder

TOO_DEEP = "nested more than 128 levels deep"

# The three builder methods that convert a Python value to JSON.
CALLS = {
    "set_space_penalty": lambda builder, value: builder.set_space_penalty(value),
    "append_character_filter": lambda builder, value: builder.append_character_filter(
        "unicode_normalize", value
    ),
    "append_token_filter": lambda builder, value: builder.append_token_filter(
        "lowercase", value
    ),
}


def nested_dicts(levels):
    """Returns {"a": {"a": ...}} made of `levels` nested dicts."""
    value = {}
    for _ in range(levels - 1):
        value = {"a": value}
    return value


def nested_lists(levels):
    """Returns {"a": [[...]]}: a dict holding `levels - 1` nested lists.

    The filter methods take a dict, so lists are wrapped in one.
    """
    value = []
    for _ in range(levels - 2):
        value = [value]
    return {"a": value}


@pytest.mark.parametrize("method", CALLS)
@pytest.mark.parametrize("make", [nested_dicts, nested_lists])
def test_accepts_128_levels(method, make):
    try:
        CALLS[method](TokenizerBuilder(), make(128))
    except ValueError as error:
        # set_space_penalty rejects the shape, but not for being too deep.
        assert TOO_DEEP not in str(error)


@pytest.mark.parametrize("method", CALLS)
@pytest.mark.parametrize("make", [nested_dicts, nested_lists])
def test_rejects_129_levels(method, make):
    with pytest.raises(ValueError, match=TOO_DEEP):
        CALLS[method](TokenizerBuilder(), make(129))


@pytest.mark.parametrize("method", CALLS)
def test_builder_stays_usable_after_rejecting_an_argument(method):
    builder = TokenizerBuilder()
    with pytest.raises(ValueError, match=TOO_DEEP):
        CALLS[method](builder, nested_dicts(129))
    CALLS[method](builder, None if method == "set_space_penalty" else {})


# Without the limit these overflowed the native stack and killed the
# interpreter, so they run in a separate process.
ISOLATED_VALUES = {
    "a value that contains itself": "value = {'rules': []}; value['self'] = value",
    "10,000 levels": textwrap.dedent(
        """
        value = {}
        for _ in range(9999):
            value = {'a': value}
        """
    ),
}


@pytest.mark.parametrize("method", CALLS)
@pytest.mark.parametrize("name", ISOLATED_VALUES)
def test_rejects_without_crashing_the_interpreter(method, name):
    call = {
        "set_space_penalty": "builder.set_space_penalty(value)",
        "append_character_filter": "builder.append_character_filter('unicode_normalize', value)",
        "append_token_filter": "builder.append_token_filter('lowercase', value)",
    }[method]
    script = "\n".join(
        [
            "from lindera import TokenizerBuilder",
            ISOLATED_VALUES[name],
            "builder = TokenizerBuilder()",
            "try:",
            f"    {call}",
            "    print('returned')",
            "except ValueError as error:",
            "    print('raised:', error)",
        ]
    )
    result = subprocess.run(
        [sys.executable, "-c", script],
        capture_output=True,
        text=True,
        timeout=60,
    )
    assert result.returncode == 0, result.stderr
    assert result.stdout.startswith("raised:"), result.stdout
    assert TOO_DEEP in result.stdout

File: lindera-ruby/test/test_argument_depth.rb
# frozen_string_literal: true

require 'rbconfig'
require_relative 'test_helper'

# The nesting limit on arguments converted to JSON (#1062).
class TestArgumentDepth < Minitest::Test
  TOO_DEEP = /nested more than 128 levels deep/
  LIB_DIR = File.expand_path('../lib', __dir__)

  # The three builder methods that convert a Ruby value to JSON.
  CALLS = {
    'set_space_penalty' => ->(builder, value) { builder.set_space_penalty(value) },
    'append_character_filter' => ->(builder, value) { builder.append_character_filter('unicode_normalize', value) },
    'append_token_filter' => ->(builder, value) { builder.append_token_filter('lowercase', value) }
  }.freeze

  # The same calls as Ruby source, for the child process.
  CALL_SOURCES = {
    'set_space_penalty' => 'builder.set_space_penalty(value)',
    'append_character_filter' => "builder.append_character_filter('unicode_normalize', value)",
    'append_token_filter' => "builder.append_token_filter('lowercase', value)"
  }.freeze

  # Without the limit these overflowed the native stack: usually a
  # SystemStackError, but some runs hung, so they run in a child process
  # with a timeout.
  ISOLATED_VALUES = {
    'a value that contains itself' => "value = { 'rules' => [] }; value['self'] = value",
    '10,000 levels' => "value = {}; 9_999.times { value = { 'a' => value } }"
  }.freeze

  # Returns { 'a' => { 'a' => ... } } made of `levels` nested hashes.
  def nested_hashes(levels)
    value = {}
    (levels - 1).times { value = { 'a' => value } }
    value
  end

  # Returns { 'a' => [[...]] }: a hash holding `levels - 1` nested arrays.
  # The filter methods take a hash, so the arrays are wrapped in one.
  def nested_arrays(levels)
    value = []
    (levels - 2).times { value = [value] }
    { 'a' => value }
  end

  # Runs `CALL_SOURCES[method]` after `value_source` in a child
  # process that is killed after `timeout` seconds.
  #
  # Returns the process status and what the child printed.
  def run_isolated(method, value_source, timeout: 30)
    script = <<~RUBY
      require 'lindera'
      #{value_source}
      builder = Lindera::TokenizerBuilder.new
      begin
        #{CALL_SOURCES.fetch(method)}
        print 'returned'
      rescue ArgumentError => e
        print "raised: \#{e.message}"
      end
    RUBY
    reader, writer = IO.pipe
    pid = Process.spawn(RbConfig.ruby, '-I', LIB_DIR, '-e', script, out: writer, err: File::NULL)
    writer.close
    deadline = Process.clock_gettime(Process::CLOCK_MONOTONIC) + timeout
    status = nil
    until status
      _, status = Process.wait2(pid, Process::WNOHANG)
      next if status

      if Process.clock_gettime(Process::CLOCK_MONOTONIC) > deadline
        Process.kill(:KILL, pid)
        _, status = Process.wait2(pid)
      else
        sleep 0.05
      end
    end
    [status, reader.read]
  ensure
    reader&.close
  end

  CALLS.each_key do |method|
    define_method("test_#{method}_accepts_128_levels") do
      [nested_hashes(128), nested_arrays(128)].each do |value|
        CALLS[method].call(Lindera::TokenizerBuilder.new, value)
      rescue StandardError => e
        # set_space_penalty rejects the shape, but not for being too deep.
        refute_match TOO_DEEP, e.message
      end
    end

    define_method("test_#{method}_rejects_129_levels") do
      [nested_hashes(129), nested_arrays(129)].each do |value|
        error = assert_raises(ArgumentError) { CALLS[method].call(Lindera::TokenizerBuilder.new, value) }
        assert_match TOO_DEEP, error.message
      end
    end

    define_method("test_#{method}_keeps_the_builder_usable") do
      builder = Lindera::TokenizerBuilder.new
      assert_raises(ArgumentError) { CALLS[method].call(builder, nested_hashes(129)) }
      CALLS[method].call(builder, method == 'set_space_penalty' ? nil : {})
    end

    ISOLATED_VALUES.each do |name, value_source|
      define_method("test_#{method}_rejects_#{name.tr(' ,', '_')}_in_a_child_process") do
        status, output = run_isolated(method, value_source)
        assert status.success?, "child exited with #{status.inspect}: #{output}"
        assert_match(/\Araised: /, output)
        assert_match TOO_DEEP, output
      end
    end
  end
end
